zebra-script/src/lib.rs

```diff
@@ -2,10 +2,155 @@
 #![doc(html_logo_url = "https://www.zfnd.org/images/zebra-icon.png")]
 #![doc(html_root_url = "https://doc.zebra.zfnd.org/zebra_script")]
 
+use displaydoc::Display;
+use thiserror::Error;
+use zcashconsensus::{
+    zcashconsensus_error_t, zcashconsensus_error_t_zcashconsensus_ERR_OK,
+    zcashconsensus_error_t_zcashconsensus_ERR_TX_DESERIALIZE,
+    zcashconsensus_error_t_zcashconsensus_ERR_TX_INDEX,
+    zcashconsensus_error_t_zcashconsensus_ERR_TX_SIZE_MISMATCH,
+};
+use zebra_chain::{
+    block::Block,
+    transaction::{TransparentInput, TransparentOutput},
+};
+
+#[derive(Debug, Display, Error)]
+#[non_exhaustive]
+pub enum Error {
+    /// Verification failed but tx was valid
+    #[non_exhaustive]
+    ErrOk,
+    /// Could not to deserialize tx
+    #[non_exhaustive]
+    TxDeserialize,
+    /// n_in is invalid for tx
+    #[non_exhaustive]
+    TxIndex,
+    /// tx is an invalid size for it's protocol
+    #[non_exhaustive]
+    TxSizeMismatch,
+    /// Encountered unknown error kind from zcashconsensus: {0}
+    #[non_exhaustive]
+    Unknown(zcashconsensus_error_t),
+}
+
+impl From<zcashconsensus_error_t> for Error {
+    #[allow(non_upper_case_globals)]
+    fn from(err_code: zcashconsensus_error_t) -> Error {
+        match err_code {
+            zcashconsensus_error_t_zcashconsensus_ERR_OK => Error::ErrOk,
+            zcashconsensus_error_t_zcashconsensus_ERR_TX_DESERIALIZE => Error::TxDeserialize,
+            zcashconsensus_error_t_zcashconsensus_ERR_TX_INDEX => Error::TxIndex,
+            zcashconsensus_error_t_zcashconsensus_ERR_TX_SIZE_MISMATCH => Error::TxSizeMismatch,
+            unknown => Error::Unknown(unknown),
+        }
+    }
+}
+
+pub fn verify_script(
+    script_pub_key: impl AsRef<[u8]>,
+    amount: i64,
+    tx_to: impl AsRef<[u8]>,
+    n_in: u32,
+    flags: u32,
+    consensus_branch_id: u32,
+) -> Result<(), Error> {
+    let script_pub_key = script_pub_key.as_ref();
+    let tx_to = tx_to.as_ref();
+
+    let script_ptr = script_pub_key.as_ptr();
+    let script_len = script_pub_key.len();
+    let tx_to_ptr = tx_to.as_ptr();
+    let tx_to_len = tx_to.len();
+    let mut err = 0;
+
+    let ret = unsafe {
+        zcashconsensus::zcashconsensus_verify_script(
+            script_ptr,
+            script_len as u32,
+            amount,
+            tx_to_ptr,
+            tx_to_len as u32,
+            n_in,
+            flags,
+            consensus_branch_id,
+            &mut err,
+        )
+    };
+
+    dbg!(&ret, &err);
+
+    if ret == 1 {
+        Ok(())
+    } else {
+        Err(Error::from(err))
+    }
+}
+pub fn script_is_valid(things: &[(zebra_chain::types::Script, TransparentOutput)]) -> bool {
+    for (script, to_spend) in things {
+        let n_in = 0;
+        let flags = 1;
+        let branch_id = 0x2bb40e61;
```

Comment:
Where do these values come from?

Reply:
these should be the `transaction.inputs().count()`, some sort of flags I don't understand, and the consensus branch id respectively. These should be changed to be args that are passed in, this was an oversight

Reply:
So `n_in` is the index in the `transaction.inputs()` that we used to find the transparentoutput being verified on that call. Each call to the zcashconsensus verify function only verifies one of the inputs in the transaction.